python/heterocl/autosa.py

```diff
@@ -0,0 +1,292 @@
+                
+                # Insert data packing and (de)serialization
+                # Create a new buffer and reshape it to original buffer after or before AutoSA func call
+                else:
+                    if arg in written_data:
+                        print(f"[ INFO ] Writing to on-chip memory {arg}. Packed into ap_uint<{target_type_bits}>...")
+                        # ALlocate new buffer and perform data deserialization
+                        deser_func = f"host_deserialize_{arg}"
+                        # Check if the size matches
+                        code = get_function_code(deser_func, header)
+                        size = get_ser_size(code)
+                        ret_code = ret_code.replace(arg, f"{arg}_sa")
+                        ret_code = f"float {arg}_sa[{size}];\n" + indent(5) + ret_code + \
+                            indent(6) + f"{deser_func}({arg}, {arg}_sa);\n"
+                    else:
+                        pass
+
+    return ret_code, header
+
+# Update HLS function names in the generated Extern IP core
+def add_prefix(header, ret_code):
+    # Preserved function keywords in AutoSA generated code
+    function_list = [
+        "autosa_func", "PE_wrapper", "PE"
+    ]
+    index = SystolicArrayRegistry.sa_module_cnt
+    for f in function_list:
+        header = header.replace(f"{f}(", f" inst{index}_{f}(")
+        ret_code = ret_code.replace(f"{f}(", f" inst{index}_{f}(")
+    SystolicArrayRegistry.sa_module_cnt += 1
+    return header, ret_code
+
+def infer_default_params(loop_bounds):
+    assert len(loop_bounds) > 1, loop_bounds
+    extra_flags = "--simd-info=./autosa_tests/mm_hcl/simd_info.json "
+    # Params for MatMul
+    if len(loop_bounds) == 3:
+        loop_bounds = [ int(_) for _ in loop_bounds ]
+        m, n, k = loop_bounds
+        if m > 1 and n > 1 and k > 1:
+            ST = 3
+            SA_dim_x = 4
+            SA_dim_y = 4
+            PART = f"{m},{n},{k}"
+            if m > 256 or n > 256 or k > 256: LAT = [16,16]
+            else: LAT = [ int(m/SA_dim_x), int(n/SA_dim_y) ]
+            LAT = [ str(1) if _ == 0 else str(_) for _ in LAT ]
+            LAT = ",".join(LAT)
+            SIMD = k if k <= 8 else 4
+        # Map reduction loop to space dim
+        else:
+            ST = 2
+            PART = "10,16"
+            LAT = "2,2"
+            SIMD = 4
+            extra_flags += "--local-reduce --reduce-op=\"+\" --simd-touch-space "
+
+    # Params for Conv
+    else:
+        OC, OH, OW, IC, R, C = loop_bounds
+        ST = 4
+        print(f"[  INFO  ] input size OC({OC}), OH({OH}), OW({OW}), IC({IC}), R({R}), C({C})")
+        PART = "16,13,13,1"
+        LAT  = "2,1,2"
+        SIMD = "1,1,2,4"
+        extra_flags = "--simd-info=./autosa_tests/cnn/simd_info.json "
+    return ST, PART, LAT, SIMD, extra_flags
+
+def generate_systolic_array(keys, values, code, backend):
```

Comment:
Seems the codegen, copying files, and generating headers are done in this function? Maybe it would be better if this function can be separated into several subfunctions or several steps like what we did in `runtime.py`.